src/services/codefixes/convertToMappedObjectType.ts

```diff
@@ -0,0 +1,99 @@
+/* @internal */
+namespace ts.codefix {
+    const fixIdAddMissingTypeof = "fixConvertToMappedObjectType";
+    const fixId = fixIdAddMissingTypeof;
+    const errorCodes = [Diagnostics.An_index_signature_parameter_type_cannot_be_a_union_type_Consider_using_a_mapped_object_type_instead.code];
+
+    type FixableDeclaration = InterfaceDeclaration | TypeAliasDeclaration;
+
+    interface Info {
+        indexSignature: IndexSignatureDeclaration;
+        container: FixableDeclaration;
+        otherMembers: ReadonlyArray<TypeElement>;
+        parameterName: Identifier;
+        parameterType: TypeNode;
+    }
+
+    registerCodeFix({
+        errorCodes,
+        getCodeActions: context => {
+            const { sourceFile, span } = context;
+            const info = getFixableSignatureAtPosition(sourceFile, span.start);
+            if (!info) return;
+            const { indexSignature, container, otherMembers, parameterName, parameterType } = info;
+
+            const changes = textChanges.ChangeTracker.with(context, t => doChange(t, sourceFile, indexSignature, container, otherMembers, parameterName, parameterType));
+            return [createCodeFixAction(fixId, changes, [Diagnostics.Convert_0_to_mapped_object_type, idText(container.name)], fixId, [Diagnostics.Convert_0_to_mapped_object_type, idText(container.name)])];
+        },
+        fixIds: [fixId],
+        getAllCodeActions: context => codeFixAll(context, errorCodes, (changes, diag) => {
+            const info = getFixableSignatureAtPosition(diag.file, diag.start);
+            if (!info) return;
+            const { indexSignature, container, otherMembers, parameterName, parameterType } = info;
+
+            doChange(changes, context.sourceFile, indexSignature, container, otherMembers, parameterName, parameterType);
+        })
+    });
+
+    function isFixableDeclaration(node: Node): node is FixableDeclaration {
+        return isInterfaceDeclaration(node) || (node.parent && isTypeLiteralNode(node) && isTypeAliasDeclaration(node.parent));
+    }
+
+    function isIndexSignatureParameterName(node: Node): node is Identifier {
+        return node && node.parent && node.parent.parent && node.parent.parent.parent &&
+            isIdentifier(node) && isParameter(node.parent) && isIndexSignatureDeclaration(node.parent.parent) && isFixableDeclaration(node.parent.parent.parent);
+    }
+
+    function getFixableSignatureAtPosition(sourceFile: SourceFile, pos: number): Info | undefined {
+        const token = getTokenAtPosition(sourceFile, pos, /*includeJsDocComment*/ false);
+        if (!isIndexSignatureParameterName(token)) return undefined;

```

Comment:
Now that Microsoft/vscode#45785 is fixed, might as well just assert these things as you go, which could save you some casts.
The only time this shouldn't be fixable is in a class, right? So in `getCodeActions`, all we should need to determine whether it's fixable is whether `token.parent.parent.parent` is a class?

Reply:
thanks, in my memory, i have not to use assert or cast 😬

Reply:
after thinking and search, This may exceed my expectations
what should i do?😢